### ISSUE TITLE
Converges because you chose a starting point directly uphill from the global optimum

### ISSUE THREAD

**Issue by JossWhittle** [CS-Swansea/Computer-Vision-and-Machine-Learning-Wiki#8]
https://github.com/CS-Swansea/Computer-Vision-and-Machine-Learning-Wiki/blame/master/code/Reading%20Group%2C%20Equation%2010.3.ipynb#L20

No free lunch theorem applies to the general case and for fitness spaces with unknown curvature and discontinuities. 

The MAE loss space is smooth in all dimensions sloping directly towards a **single** global optimum, so convergence is guaranteed in the limit for sensibly chosen learning rates. 

If you add a constraint to the loss function which introduces local optima and sharp edges, such as:

```
# Local minima
red_herring_probability = np.array([1.0, 1.0, 0.0, 1.0, 0.0])
red_herring_mae = max(np.sum(np.abs(red_herring_probability - person)), 0.5)

# Global minima
true_mae = np.sum(np.abs(true_probability - person))

# Constraint that adds some cliff edges
constraint = (1 if np.sum(person) == np.sum(true_probability) else np.inf)

fitness = min(true_mae, red_herring_mae) * constraint
```

Which has the same global optima, then the algorithm is not guaranteed to converge, especially when the size of the genome is increased from 5 to something like 100. 

In real world problems for GA you don't have the `true_probability` to make a direct comparison to, you have some task or objective function defined `fitness(person)`, which can have much more interesting properties (and disjoint regions).

**Comment by JossWhittle** [comment]
@MPKenning :p

**Comment by medwar99** [comment]
> @MPKenning :p

I cannot get this minimal example to run, can you check ":p" and provide a complete example?

**Comment by MPKenning** [comment]
I have added Joss' comments to enrich the Notebook.
